### PR TITLE
[SPARK-47654][INFRA] Structured logging framework: support log concatenation

### PR DESCRIPTION
**Pull request by gengliangwang**
### What changes were proposed in this pull request?

Support the log concatenation in the structured logging framework. For example
```
log"${MDC(CONFIG, SHUFFLE_MAPOUTPUT_MIN_SIZE_FOR_BROADCAST.key)} " +
  log"(${MDC(MIN_SIZE, minSizeForBroadcast.toString)} bytes) " +
  log"must be <= spark.rpc.message.maxSize (${MDC(MAX_SIZE, maxRpcMessageSize.toString)} " +
  log"bytes) to prevent sending an rpc message that is too large." 
```

### Why are the changes needed?

Although most of the Spark logs are short, we need this convenient syntax when handling long logs with multiple variables.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

New UT

### Was this patch authored or co-authored using generative AI tooling?

Yes, GitHub copilot

### CODE REVIEW

**Comment by gengliangwang** [comment]
After this one, we can kick off the log migration tasks under https://issues.apache.org/jira/browse/SPARK-47240

File: core/src/main/scala/org/apache/spark/MapOutputTracker.scala [hunk 1]
```diff
@@ -725,11 +726,13 @@ private[spark] class MapOutputTrackerMaster(
...
-      s"must be <= spark.rpc.message.maxSize ($maxRpcMessageSize bytes) to prevent sending an " +
-      "rpc message that is too large."
+    val msg = log"${MDC(CONFIG, SHUFFLE_MAPOUTPUT_MIN_SIZE_FOR_BROADCAST.key)} " +
+      log"(${MDC(MIN_SIZE, minSizeForBroadcast.toString)} bytes) " +
+      log"must be <= spark.rpc.message.maxSize (${MDC(MAX_SIZE, maxRpcMessageSize.toString)} " +
+      log"bytes) to prevent sending an rpc message that is too large."
+
     logError(msg)
-    throw new IllegalArgumentException(msg)
+    throw new IllegalArgumentException(msg.message)
```
Comment (dongjoon-hyun) [inline core/src/main/scala/org/apache/spark/MapOutputTracker.scala:735]: This seems to be touched as the example. Shall we revert this?

Comment (gengliangwang) [reply core/src/main/scala/org/apache/spark/MapOutputTracker.scala]: Sure, reverted

**Review by cloud-fan — APPROVED** [review]

**Comment by gengliangwang** [comment]
Thanks, merging to master

**Pull request closed without merging**